python/ql/lib/semmle/python/frameworks/Pydantic.qll

```diff
@@ -0,0 +1,101 @@
+ *
+ * See
+ * - https://pypi.org/project/pydantic/
+ * - https://pydantic-docs.helpmanual.io/
+ */
+
+private import python
+private import semmle.python.dataflow.new.DataFlow
+private import semmle.python.dataflow.new.TaintTracking
+private import semmle.python.Concepts
+private import semmle.python.ApiGraphs
+
+/**
+ * INTERNAL: Do not use.
+ *
+ * Provides models for `pydantic` PyPI package.
+ *
+ * See
+ * - https://pypi.org/project/pydantic/
+ * - https://pydantic-docs.helpmanual.io/
+ */
+module Pydantic {
+  /**
+   * Provides models for `pydantic.BaseModel` subclasses (a pydantic model).
+   *
+   * See https://pydantic-docs.helpmanual.io/usage/models/.
+   */
+  module BaseModel {
+    /** Gets a reference to a `pydantic.BaseModel` subclass (a pydantic model). */
+    API::Node subclassRef() {
+      result = API::moduleImport("pydantic").getMember("BaseModel").getASubclass+()
+    }
+
+    /**
+     * A source of instances of `pydantic.BaseModel` subclasses, extend this class to model new instances.
+     *
+     * This can include instantiations of the class, return values from function
+     * calls, or a special parameter that will be set when functions are called by an external
+     * library.
+     *
+     * Use the predicate `BaseModel::instance()` to get references to instances of `pydantic.BaseModel`.
+     */
+    abstract class InstanceSource extends DataFlow::LocalSourceNode { }
+
+    /** Gets a reference to an instance of a `pydantic.BaseModel` subclass. */
+    private DataFlow::TypeTrackingNode instance(DataFlow::TypeTracker t) {
+      t.start() and
+      result instanceof InstanceSource
+      or
+      t.start() and
+      instanceStepToPydanticModel(_, result)
+      or
+      exists(DataFlow::TypeTracker t2 | result = instance(t2).track(t2, t))
+    }
+
+    /** Gets a reference to an instance of a `pydantic.BaseModel` subclass. */
+    DataFlow::Node instance() { instance(DataFlow::TypeTracker::end()).flowsTo(result) }
+
+    /**
+     * A step from an instance of a `pydantic.BaseModel` subclass, that might result in
+     * an instance of a `pydantic.BaseModel` subclass.
+     *
+     * NOTE: We currently overapproximate, and treat all attributes as containing another
+     * pydantic model. For the code below, we _could_ limit this to `main_foo` and
+     * members of `other_foos`.
+     *
+     * ```py
+     * class MyComplexModel(BaseModel):
+     * field: str
+     * main_foo: Foo
+     * other_foos: List[Foo]
+     * ```
+     */
+    private predicate instanceStepToPydanticModel(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+      // attributes (such as `model.foo`)
+      nodeFrom = instance() and
+      nodeTo.(DataFlow::AttrRead).getObject() = nodeFrom
+      or
+      // subscripts on attributes (such as `model.foo[0]`)
+      nodeFrom.(DataFlow::AttrRead).getObject() = instance() and
+      nodeTo.asCfgNode().(SubscriptNode).getObject() = nodeFrom.asCfgNode()
```

Comment:
```suggestion
      // subscripts (such as `model[0]`)
      nodeFrom = instance() and
      nodeTo.asCfgNode().(SubscriptNode).getObject() = nodeFrom.asCfgNode()
```
I believe this will also cover cases such as `further_foos: List[List[Foo]]` which would otherwise be unreachable.

Reply:
Good point :+1: